File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        int n=nums.size();
        vector<vector<int>> res;
        for(int i=0;i<n;i++){
            if(nums[i]>0) break;
            if(i>0 && nums[i]==nums[i-1]) continue;
            int l=i+1,r=n-1;
            while(l<r){
                int sum=nums[i]+nums[l]+nums[r];
                if(sum>0) r--;
                else if(sum<0) l++;
                else{
                    res.push_back({nums[i],nums[l],nums[r]});
                    int x=nums[l],y=nums[r];
                    while(l<r && nums[l]==x) l++;
                    while(l<r && nums[r]==y) r--;
                }
            }
        }
        return res;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        res = set()
        for i,num in enumerate(nums):
            if num>0: break
            if i>0 and nums[i-1]==num:
                continue
            if num>0:
                break
            j,k=i+1,n-1
            while j<k:
                if nums[i]+nums[j]+nums[k] == 0:
                    res.add((nums[i],nums[j],nums[k]))
                    j+=1
                elif nums[i]+nums[j]+nums[k] < 0:
                    j+=1
                else:
                    k-=1
        return list(res)
